libmattermost.c

```diff
@@ -1322,36 +1391,155 @@ mm_get_teams(MattermostAccount *ma)
 	g_free(url);
 	
 }
+
+static void 
+mm_save_user_pref_response(MattermostAccount *ma, JsonNode *node, gpointer user_data)
+{
+	MattermostUserPref *pref = user_data;
+	g_free(pref);
+	
+	if (json_node_get_node_type(node) == JSON_NODE_OBJECT) {
+		JsonObject *response = json_node_get_object(node);
+		if (json_object_get_int_member(response, "status_code") >= 400) {
+			purple_notify_error(ma->pc, _("Save Preferences Error"), _("There was an error saving user preferences"), json_object_get_string_member(response, "message"), purple_request_cpar_from_connection(ma->pc));
+		return;
+        }
+	}
+}
+
+static void
+mm_save_user_pref(MattermostAccount *ma, MattermostUserPref *pref)
+{
+	JsonArray *data = json_array_new();
+	JsonObject *pref_data = json_object_new();
+	gchar *postdata, *url;
+	
+	json_object_set_string_member(pref_data, "user_id", pref->user_id);
+	json_object_set_string_member(pref_data, "category", pref->category);
+	json_object_set_string_member(pref_data, "name", pref->name);
+	json_object_set_string_member(pref_data, "value", pref->value);
+
+	json_array_add_object_element(data,pref_data);
+	postdata = json_array_to_string(data);
+	
+	if (!g_strcmp0(pref->category,"direct_channel_show")) {
+		url = mm_build_url(ma, "/api/v3/preferences/save");
+		mm_fetch_url(ma, url, postdata, mm_save_user_pref_response, pref);
+	}
+
+	g_free(postdata);
+	json_array_unref(data);
+}
+
+int
+mm_compare_prefs_int(gconstpointer a, gconstpointer b)
+{
+	const MattermostUserPref *p1 = a;
+	const MattermostUserPref *p2 = b;
+
+	if (!(g_strcmp0(p1->user_id,p2->user_id) &&

```

Comment:
Not sure whether this block should get `purple_strequal()`s or whether it should be doing individual compares and returning -1/1/0 accordingly.  (Have to be careful with compare functions in glib that only return 0/-1 or 0/1 as it can cause infinite loops when sorting, but that won't affect things here for a search)